File: api/src/main/java/xtdb/api/IXtdb.java
package xtdb.api;

import xtdb.query.Query;
import xtdb.query.QueryOpts;
import xtdb.tx.Ops;

import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.stream.Stream;

public interface IXtdb extends AutoCloseable {

    CompletableFuture<Stream<Map<String, ?>>> openQueryAsync(Query q, QueryOpts opts);

    private static <T> T await(CompletableFuture<T> fut) {
        try {
            return fut.get();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        } catch (ExecutionException e) {
            throw new RuntimeException(e.getCause());
        }
    }

    default CompletableFuture<Stream<Map<String, ?>>> openQueryAsync(Query q) {
        return openQueryAsync(q, new QueryOpts());
    }

    default Stream<Map<String, ?>> openQuery(Query q) {
        return await(openQueryAsync(q));
    }

    default Stream<Map<String, ?>> openQuery(Query q, QueryOpts opts) {
        return await(openQueryAsync(q, opts));
    }

    CompletableFuture<TransactionKey> submitTxAsync(List<Ops> ops, TxOptions txOpts);

    default CompletableFuture<TransactionKey> submitTxAsync(List<Ops> ops) {
        return submitTxAsync(ops, new TxOptions());
    }

    default TransactionKey submitTx(List<Ops> ops, TxOptions txOpts) {
        return await(submitTxAsync(ops, txOpts));
    }

    default TransactionKey submitTx(List<Ops> ops) {
        return await(submitTxAsync(ops));
    }

    @Override
    void close();
}

File: api/src/main/kotlin/xtdb/api/IXtdb.kt
package xtdb.api

import xtdb.api.query.QueryOptions
import xtdb.api.query.XtqlQuery
import xtdb.api.tx.TxOp
import xtdb.api.tx.TxOptions
import java.util.stream.Stream

@Suppress("OVERLOADS_INTERFACE")
interface IXtdb : AutoCloseable {

    /**
     * Opens an XTQL query - see [XtqlQuery] for more details on XTQL.
     *
     * By default, this method will block indefinitely until the node has indexed the [requested transaction][QueryOptions.afterTx], or
     * you can specify a [timeout][QueryOptions.txTimeout].
     *
     * @param xtql the XTQL query
     * @param opts query options
     * @return the results stream.
     *         This result stream MUST be explicitly closed when no longer required.
     */
    @JvmOverloads
    fun openQuery(xtql: XtqlQuery, opts: QueryOptions = QueryOptions()) : Stream<Map<String, *>>

    /**
     * Opens an SQL query - see the [SQL documentation](https://docs.xtdb.com/reference/main/sql/queries) for more details on XTDB's SQL support.
     *
     * By default, this method will block indefinitely until the node has indexed the [requested transaction][QueryOptions.afterTx], or
     * you can specify a [timeout][QueryOptions.txTimeout].
     *
     * @see openQuery
     * @param sql the SQL query
     * @param opts query options
     * @return the results stream.
     *         This result stream MUST be explicitly closed when no longer required.
     */
    @JvmOverloads
    fun openQuery(sql: String, opts: QueryOptions = QueryOptions()) : Stream<Map<String, *>>

    /**
     * Submits transactions to the log for processing - this method will block
     * until the log has confirmed receipt of the transaction.
     *
     * @param txOpts options for the transaction
     * @param ops XTQL/SQL transaction operations.
     * @return the transaction key of the submitted transaction.
     */
    fun submitTx(txOpts: TxOptions, vararg ops: TxOp): TransactionKey

    /**
     * Submits transactions to the log for processing - this method will block
     * until the log has confirmed receipt of the transaction.
     *
     * @param ops XTQL/SQL transaction operations.
     * @return the transaction key of the submitted transaction.
     */
    fun submitTx(vararg ops: TxOp) = submitTx(TxOptions(), *ops)

    /**
     * Executes the transaction - this method will block until the receiving node has indexed the transaction.
     *
     * @param txOpts options for the transaction
     * @param ops XTQL/SQL transaction operations.
     * @return the result of the executed transaction.
     */
    fun executeTx(txOpts: TxOptions, vararg ops: TxOp): TransactionResult

    /**
     * Executes the transaction - this method will block until the receiving node has indexed the transaction.
     *
     * @param ops XTQL/SQL transaction operations.
     * @return the result of the executed transaction.
     */
    fun executeTx(vararg ops: TxOp) = executeTx(TxOptions(), *ops)

    /**
     * @suppress
     */
    override fun close()
}
